clients/unshaded/src/main/java/tachyon/client/ResourcePool.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.client;
+
+import java.io.IOException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * Class representing a pool of resources to be temporarily used and returned. Inheriting classes
+ * should implement the close method as well as initialize the resources in the constructor.
+ *
+ * @param <T> The type of resource this pool manages.
+ */
+public abstract class ResourcePool<T> {
+  protected final Object mCapacityLock;
+  protected int mCurrentCapacity;
+  protected final int mMaxCapacity;
+  protected BlockingQueue<T> mResources;
+
+  public ResourcePool(int maxCapacity) {
+    mCapacityLock = new Object();
+    mMaxCapacity = maxCapacity;
+    mResources = new LinkedBlockingQueue<T>(maxCapacity);
+  }
+
+  /**
+   * Acquires an object of type T, this operation is blocking if no clients are available.
+   *
+   * @return a MasterClient, guaranteed to be only available to the caller
+   */
+  public T acquire() {
+    // If the resource pool is empty but capacity is not yet full, create a new resource.
+    synchronized (mCapacityLock) {
+      if (mResources.isEmpty() && mCurrentCapacity < mMaxCapacity) {
+        T newResource = createNewResource();
+        mCurrentCapacity ++;
+        return newResource;
```

Comment:
So no one is tracking this newly created `newResource`? What if resource consumer fails to release resource? I think it is probably better to at least have a list containing all acquired resources.